Add update agent example application

gabriela-yoncheva opened the pull request:
[#4] Add update agent example application

Created update agent example application similar to the containers update agent:
- has Dockerfile similar to the vehicle-simulator application
- uses the update agent API
- is used for managing files in a directory, message to feature contains filenames and URLs

gabriela-yoncheva committed:
[#4] Add update agent example application

Created update agent example application similar to the containers update agent:

- has Dockerfile similar to the vehicle-simulator application

- uses the update agent API

- is used for managine files in a directory, message to feature contains filenames and URLs

Signed-off-by: Gabriela Yoncheva <Gabriela.Yoncheva2@bosch.com>

k-gostev commented on custom-update-agent/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+RUN GOOS=$TARGETOS GOARCH=$TARGETARCH go build -o app .
+
+FROM scratch
+COPY --from=builder /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/
```
This will only work for Debians and may cause confusion if someone try this on a different linux OS. Ref: https://go.dev/src/crypto/x509/root_linux.go

Instead of this we can either skip security verification from the code (which is sort of ok for example apps), we can use container mounts, or can just add description here in the comment what to change. @dimitar-dimitrow wdyt?

dimitar-dimitrow replied:
In the current implementation the default go http Client is used. Do you know whether it performs any security verification? My preferences is for a mount, this way the user can choose to provide a non system CA. However this mounting should be described in a README which is currently missing.

k-gostev replied:
As far as I remember (but don't quote me on this) it tries to verify the server certificate whether it is trusted or not. And for this it uses what is available [here](https://go.dev/src/crypto/x509/root_linux.go) as a sort of trust store. I guess this is why @gabriela-yoncheva added this `COPY` line there, because with the alpine base image there is nothing (or almost nothing), making any HTTPS downloads impossible.

k-gostev replied:
Adding a mount point for this would complicate a bit the deployment of this application. The README should include two different `run` parts, one for building and starting as a standard application and second part for containerisation of the app.

Another possible solution for this problem might be to use different base platform layer.

k-gostev commented on custom-update-agent/go.mod:
```diff
@@ -0,0 +1,23 @@
+module github.com/eclipse-kanto/example-applications/custom-update-agent
+
+go 1.18
```
```suggestion
go 1.21
```

k-gostev commented on custom-update-agent/main.go:
```diff
@@ -0,0 +1,93 @@
+)
+
+const (
+
+	// default local connection config
```
```suggestion
```

k-gostev commented on custom-update-agent/main.go:
```diff
@@ -0,0 +1,93 @@
+	subscribeTimeoutDefault            = "15s"
+	unsubscribeTimeoutDefault          = "5s"
+
+	// default update agent config
```
```suggestion
```

k-gostev commented on custom-update-agent/main.go:
```diff
@@ -0,0 +1,93 @@
+}
+
+func main() {
+
+	//set logger level and output
```
```suggestion
```

k-gostev commented on custom-update-agent/main.go:
```diff
@@ -0,0 +1,93 @@
+	sig := <-signalChan
+	slog.Info("Exiting!, recieved", "signal", sig)
+	updateAgent.(api.UpdateAgent).Stop()
+
```
```suggestion
```

k-gostev commented on custom-update-agent/updateagent/update_manager.go:
```diff
@@ -0,0 +1,192 @@
+
+// WatchEvents subscribes for events that update the current state inventory
+func (updMgr *fileUpdateManager) WatchEvents(ctx context.Context) {
+	// no container events handled yet - current state inventory reported only on initial start or explicit get request
```
Copy-paste? I think you can add just some description what should be implemented in that function.

k-gostev commented on custom-update-agent/updateagent/update_operation.go:
```diff
@@ -0,0 +1,518 @@
+
+// Identify executes the IDENTIFYING phase, triggered with the full desired state for the domain
+func (o *operation) Identify() (bool, error) {
+
```
```suggestion
```

k-gostev commented on custom-update-agent/updateagent/update_operation.go:
```diff
@@ -0,0 +1,518 @@
+// ActionRemove: removes the old file from fileagent directory.
+// ActionAdd and ActionReplace: removes temporary download directory.
+func cleanup(o *operation, baselineAction *action) {
+
```
```suggestion
```

k-gostev commented on custom-update-agent/updateagent/update_operation.go:
```diff
@@ -0,0 +1,518 @@
+}
+
+func (o *operation) addFile(desired *util.File) error {
+
```
```suggestion
```

k-gostev commented on custom-update-agent/updateagent/update_operation.go:
```diff
@@ -0,0 +1,518 @@
+	err := os.Remove("./fileagent/" + desired.Name)
+	if err != nil {
+		slog.Error(fmt.Sprintf("got error removing file [%s]", desired.Name), "error", err)
+
```
```suggestion
```

k-gostev commented on custom-update-agent/updateagent/update_operation.go:
```diff
@@ -0,0 +1,518 @@
+		slog.Error("got error removing temporary folders", "error", err)
+	}
+	return err
+}
+func (o *operation) removeFile(desired *util.File) error {
```
```suggestion
}

func (o *operation) removeFile(desired *util.File) error {
```

k-gostev commented on custom-update-agent/updateagent/update_operation.go:
```diff
@@ -0,0 +1,518 @@
+
+	}
+	return err
+}
+func (o *operation) downloadFile(desired *util.File) error {
```
```suggestion
}

func (o *operation) downloadFile(desired *util.File) error {
```

k-gostev commented:
It appears that one of my comments is missing somewhere. So, besides what we are already discussing, this PR needs:
1. README file with description of the functionality and an installation guide.
2. Github workflow addition
3. deployment.json file like [this](https://github.com/eclipse-kanto/example-applications/blob/main/vehicle-simulator/deployment.json)
4. golint, go fmt, go vet, to be executed.

dimitar-dimitrow commented on custom-update-agent/go.mod:
```diff
@@ -0,0 +1,23 @@
+
+require (
+	github.com/eclipse-kanto/container-management v0.1.0-M4
+	github.com/eclipse-kanto/update-manager v0.0.0-20230628072101-b91f0c30e00f
```
Use the latest update-manager version, as the 0.0.0-20230628072101-b91f0c30e00f points to the initial contribution commit. Afterwards there are some UA API changes that should be reflected in the code.

dimitar-dimitrow commented on custom-update-agent/main.go:
```diff
@@ -0,0 +1,93 @@
+	unsubscribeTimeoutDefault          = "5s"
+
+	// default update agent config
+	updateAgentEnableDefault                 = false
```
Could be removed as it is not used.
```suggestion
```

dimitar-dimitrow commented on custom-update-agent/main.go:
```diff
@@ -0,0 +1,93 @@
+	// default update agent config
+	updateAgentEnableDefault                 = false
+	updateAgentDomainDefault                 = "files"
+	updateAgentVerboseInventoryReportDefault = false
```
Could be removed as it is not used.
```suggestion
```

dimitar-dimitrow commented on custom-update-agent/main.go:
```diff
@@ -0,0 +1,93 @@
+const (
+
+	// default local connection config
+	connectionBrokerURLDefault         = "tcp://localhost:1883"
+	connectionKeepAliveDefault         = "20s"
+	connectionDisconnectTimeoutDefault = "250ms"
+	connectionClientUsername           = ""
+	connectionClientPassword           = ""
+	connectTimeoutTimeoutDefault       = "30s"
+	acknowledgeTimeoutDefault          = "15s"
+	subscribeTimeoutDefault            = "15s"
+	unsubscribeTimeoutDefault          = "5s"
```
`mqtt.NewDefaultConfig()` could be used from the update manager.

dimitar-dimitrow commented on custom-update-agent/updateagent/update_agent_opts.go:
```diff
@@ -0,0 +1,122 @@
+// FileUpdateAgentOpt represents the available configuration options for the Containers UpdateAgent service
+type FileUpdateAgentOpt func(updateAgentOptions *updateAgentOpts) error
+
+type updateAgentOpts struct {
```
This file and struct seems redundant as the updateAgentOpts consist of domain + mqtt.ConnectionConfig. So you can directly use those two parameters(domain+connectionConfig) for UpdateAgentClient and UpdateAgent creation. I know that this way of implementation comes from the Container's Update Agent, but as it is an example application, it would be better to keep it as simple as possible.

gabriela-yoncheva committed:
[#4] Add update agent example application

Changes added:
- all requested changes have been applied
- mount points are used both for adding needed certificates and for specifying the directory to be managed by the update agent
- deployment.json provided and tested with the -f flag used by kanto-cm
- state.props file will be created if not avaliable in managed directory
- backup functionality for rollback is now working
- readme.md provided with basic information about the update agent and instruction on how to install it (standard/containerized)

Note:
- could not push changes to validation.yaml, it still need to be updated!
Signed-off-by: Gabriela Yoncheva <Gabriela.Yoncheva2@bosch.com>

dimitar-dimitrow commented on custom-update-agent/updateagent/update_agent_init.go:
```diff
@@ -0,0 +1,73 @@
+	mqttClient, _ := mqtt.NewUpdateAgentClient(domainName, &mqtt.ConnectionConfig{
+		Broker:             broker,
+		KeepAlive:          keepAlive,
+		DisconnectTimeout:  disconnectTimeout,
+		Username:           clientUsername,
+		Password:           clientPassword,
+		ConnectTimeout:     connectTimeout,
+		AcknowledgeTimeout: acknowledgeTimeout,
+		SubscribeTimeout:   subscribeTimeout,
+		UnsubscribeTimeout: unsubscribeTimeout,
+	})
+
+	return agent.NewUpdateAgent(mqttClient, newUpdateManager(domainName)), nil
+}
+
+// newUpdateManager instantiates a new update manager instance
+func newUpdateManager(domainName string) api.UpdateManager {
+	return &fileUpdateManager{
+		domainName: domainName,
+
+		createUpdateOperation: newOperation,
+	}
+}
+
+// Init initializes a new Update Agent instance using given configuration and domain
+func Init(config mqtt.ConnectionConfig, domain string) (interface{}, error) {
+
+	return newUpdateAgent(
+		domain,
+		config.Broker,
+		config.KeepAlive,
+		config.DisconnectTimeout,
+		config.Username,
+		config.Password,
+		config.ConnectTimeout,
+		config.AcknowledgeTimeout,
+		config.SubscribeTimeout,
+		config.UnsubscribeTimeout,
+	)
+}
```
Could be simplified.
```suggestion

// newUpdateManager instantiates a new update manager instance
func newUpdateManager(domainName string) api.UpdateManager {
	return &fileUpdateManager{
		domainName: domainName,

		createUpdateOperation: newOperation,
	}
}

// Init initializes a new Update Agent instance using given configuration and domain
func Init(config *mqtt.ConnectionConfig, domain string) (interface{}, error) {
	mqttClient, err := mqtt.NewUpdateAgentClient(domainName, config)
	if err != nil {
	    return nil, err
	}

	return agent.NewUpdateAgent(mqttClient, newUpdateManager(domainName)), nil
}
```

dimitar-dimitrow commented on custom-update-agent/main.go:
```diff
@@ -0,0 +1,53 @@
+	flag.Parse()
+	updateagent.FileDirectory = *fileDirPtr
+
+	updateAgent, _ := updateagent.Init(*mqtt.NewDefaultConfig(), "files")
```
Check the error and exit with an error code.
```suggestion
	updateAgent, err := updateagent.Init(*mqtt.NewDefaultConfig(), "files")
	if err != nil {
	    slog.Error("could not initialize an Update Agent service! got", "error", err)
	    os.Exit(1)
	}
```

dimitar-dimitrow commented on custom-update-agent/main.go:
```diff
@@ -0,0 +1,53 @@
+
+	updateAgent, _ := updateagent.Init(*mqtt.NewDefaultConfig(), "files")
+	err := updateAgent.(api.UpdateAgent).Start(context.Background())
+	if err != nil {
+		slog.Error("could not start Update Agent service! got", "error", err)
+	} else {
+		slog.Info("successfully started Update Agent service")
+	}
```
Upon error exit with an error code.
```suggestion
	if err != nil {
		slog.Error("could not start Update Agent service! got", "error", err)
		os.Exit(1)
	}
	slog.Info("successfully started Update Agent service")
```

dimitar-dimitrow commented on custom-update-agent/README.md:
```diff
@@ -0,0 +1,139 @@
+* Eclipse Kanto Container Management
+
+
+Regardless if you are running the update agent as a standard or containerized application, you will need to add the following to the update manager configuration, located at `/etc/update-manager/config.json`:
```
```suggestion
Regardless if you are running the update agent as a standard or containerized application, you will need to add the following to the update manager configuration, located at `/etc/kanto-update-manager/config.json`:
```

dimitar-dimitrow commented on custom-update-agent/util/to_files.go:
```diff
@@ -0,0 +1,40 @@
+}
+
+func toFile(component *types.ComponentWithConfig) (*File, error) {
+	file := &File{
+		Name:        component.Config[0].Value,
+		DownloadURL: component.Config[1].Value,
+	}
```
Do not rely on the order of the configurations, as the DownloadURL could be assigned with the file_name. Use the keys to assign the proper fields.
```suggestion
	file := &File{}
	for _, kvPair := range component.Config {
		if kvPair.Key == "file_name" {
			file.Name = kvPair.Value
		}
		if kvPair.Key == "download_url" {
			file.DownloadURL = kvPair.Value
		}
	}
```

dimitar-dimitrow reviewed (changes requested):
The custom-update-agent should be included in the [validation workflow](https://github.com/eclipse-kanto/example-applications/blob/main/.github/workflows/validation.yaml).

daniel-milchev commented on custom-update-agent/updateagent/update_manager.go:
```diff
@@ -0,0 +1,216 @@
+
+// Name returns the name of this update manager, e.g. "files".
+func (updMgr *fileUpdateManager) Name() string {
+
```
```suggestion

```

daniel-milchev commented on custom-update-agent/updateagent/update_manager.go:
```diff
@@ -0,0 +1,216 @@
+	if err != nil {
+		slog.Error("got error checking current files", "error", err)
+		return nil
+
```
```suggestion

```

daniel-milchev commented on custom-update-agent/updateagent/update_manager.go:
```diff
@@ -0,0 +1,216 @@
+		if err != nil {
+			slog.Error("got error checking current files", "error", err)
+			return nil
+
```
```suggestion

```

daniel-milchev commented on custom-update-agent/updateagent/update_manager.go:
```diff
@@ -0,0 +1,216 @@
+		_, err = os.Create(propsFilePath)
+		if err != nil {
+			slog.Error(fmt.Sprintf("got error creating file [%s]", "state.props"), "error", err)
+
```
```suggestion

```

daniel-milchev commented on custom-update-agent/updateagent/update_operation.go:
```diff
@@ -0,0 +1,587 @@
+func (o *operation) Execute(command types.CommandType, baseline string) {
+	commandHandler, action := o.getCommandHandler(baseline, command)
+	if action == nil {
+
```
```suggestion

```

daniel-milchev commented on custom-update-agent/updateagent/update_operation.go:
```diff
@@ -0,0 +1,587 @@
+
+		return
+	}
+
```
```suggestion

```

daniel-milchev commented on custom-update-agent/updateagent/update_operation.go:
```diff
@@ -0,0 +1,587 @@
+
+			err := addProperty(action.desired.Name, action.desired.DownloadURL)
+
+			if err != nil {
```
```suggestion
			if err := addProperty(action.desired.Name, action.desired.DownloadURL); err != nil {
```

daniel-milchev commented on custom-update-agent/updateagent/update_operation.go:
```diff
@@ -0,0 +1,587 @@
+			if err != nil {
+				lastActionErr = err
+				return
+
```
```suggestion

```

daniel-milchev commented on custom-update-agent/updateagent/update_operation.go:
```diff
@@ -0,0 +1,587 @@
+			}
+			lastActionMessage = "File added to directory."
+		} else if action.actionType == util.ActionRemove {
+			err := o.removeFile(action.current)
+
+			if err != nil {
```
```suggestion
			if err := o.removeFile(action.current); err != nil {
```

daniel-milchev commented on custom-update-agent/updateagent/update_operation.go:
```diff
@@ -0,0 +1,587 @@
+	if err != nil {
+		slog.Error("got error reading state.props file", "error", err)
+		return false, err
+
```
```suggestion

```

daniel-milchev commented on custom-update-agent/updateagent/update_operation.go:
```diff
@@ -0,0 +1,587 @@
+	if err != nil {
+		slog.Error("got error opening state.props file", "error", err)
+		return false, err
+
```
```suggestion

```

daniel-milchev commented on custom-update-agent/updateagent/update_manager.go:
```diff
@@ -0,0 +1,216 @@
+
+		slog.Error("got error when reading state.props file", "error", err)
+		return nil
+
```
```suggestion

```

daniel-milchev commented on custom-update-agent/updateagent/update_manager.go:
```diff
@@ -0,0 +1,216 @@
+
+	properties, err := props.Read(propsFile)
+	if err != nil {
+
```
```suggestion

```

k-gostev committed:
Addressing comments

Signed-off-by: Kristiyan Gostev <kristiyan.gostev@bosch.com>
